modules/mcaptcha.nix

```diff
@@ -0,0 +1,279 @@
+{
+  config,
+  lib,
+  options,
+  pkgs,
+  ...
+}:
+with builtins;
+with lib; let
+  cfg = config.services.mcaptcha;
+
+  settingsFormat = pkgs.formats.toml {};
+  filteredSettings = lib.attrsets.filterAttrsRecursive (_path: value: value != null) cfg.settings;
+  configFile = settingsFormat.generate "mcaptcha.config.toml" filteredSettings;
+in {
+  options.services.mcaptcha.enable = mkEnableOption "Enable mCaptcha server.";
+  options.services.mcaptcha.package = mkPackageOption pkgs "mcaptcha" {};
+
+  options.services.mcaptcha.settings = lib.mkOption {
+    type = lib.types.submodule {
+      freeformType = settingsFormat.type;
+
+      options.database.name = mkOption {
```

Comment:
Consider moving all these options into an attribute set:

```nix
options = {
  database.name = mkOption { /* ... */ };
  database.username = mkOption { /* ... */ };
  /* ... */
};
```

Reply:
Unresolved.

Reply:
Personally, I like this Nix language feature and use it pretty much wherever I can. I bet if more tried it more would like it, as well. It's there for a reason.

I like two aspects of this feature:

In cases of long literal attribute sets it makes the full path of the value clear from any line. No scrolling required.

Text diffing (what we have in git) is more agreeable with this.
For example, changing from

```nix
{
 foo.bar = 2;  
};
```

Could be either:
```nix
{
  foo = {
    bar = 2;
    baz = 3;
  }
}
```

Or:
```nix
{
  foo.bar = 2;
  foo.baz = 3;
}
```

And so on.

Is this convincing?